src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+            _ if *v == [] => StackElement::Dissatisfied,
+            _ => StackElement::Push(v)
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HashLockType<'desc>{
+    ///SHA 256 hashlock
+    Sha256HashLock(&'desc sha256::Hash),
+    ///Hash 256 hashlock
+    Hash256HashLock(&'desc sha256d::Hash),
+    ///Hash160 hashlock
+    Hash160HashLock(&'desc hash160::Hash),
+    ///Ripemd160 hashlock
+    Ripemd160HashLock(&'desc ripemd160::Hash),
+}
+/// A satisfied Miniscript condition (Signature, Hashlock, Timelock)
+/// 'desc represents the lifetime of descriptor and `stack represents the lifetime of witness
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum SatisfiedConstraint<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    ///Public key and corresponding signature
+    PublicKey {
+        key: &'desc Pk,
+        sig: secp256k1::Signature,
+    },
+    ///Public keyhash, corresponding pubkey and signature
+    PublicKeyHasH {
+        keyhash: &'desc Pkh,
+        key: bitcoin::PublicKey,
+        sig: secp256k1::Signature,
+    },
+    ///Hashlock and preimage for SHA256
+    HashLock {
+        hash: HashLockType<'desc>,
+        preimage: &'stack [u8],
+    },
+    ///Relative Timelock for CSV.
+    RelativeTimeLock {
+        time: &'desc u32,
+    },
+    ///Absolute Timelock for CLTV.
+    AbsoluteTimeLock {
+        time: &'desc u32,
+    },
+}
+
+///This is used by the interpreter to know which evaluation state a AstemElem is.
+///This is required because whenever a same node(for eg. OrB) appears on the stack, we don't
+///know if the left child has been evaluated or not. And based on the result on
+///the top of the stack, we need to decide whether to execute right child or not.
+///This is also useful for wrappers and thresholds which push a value on the stack
+///depending on evaluation of the children.
+enum NodeEvaluationState{
+    ///No children have been evaluted
+    NoChildrenCompleted,
+    ///Only the left child has been evaluated
+    LeftChildCompleted,
+    ///Only partial evaluation completed(used for Thresh nodes)
+    PartialThreshCompleted,
+    ///Multi-sig verification which represents currently k out n are satisfied.
+    ///This is required because the children of multi are not AstElem and
+    ///cannot be interpreted directly. Thus, we need to keep track of them
+    ///separately
+    PartialMultiCompleted(usize,usize),
+    ///All chilren have been evaluated
+    AllChildrenCompleted
+}
+
+/// An iterator over all the constraints satisfied by a given
+/// descriptor/scriptSig/witness stack tuple
+pub struct SatisfiedConstraints<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    secp: &'desc secp256k1::Secp256k1<VerifyOnly>,
+    sighash: &'desc secp256k1::Message,
+    public_key: Option<&'desc Pk>,
+    state : Vec <(&'desc AstElem<Pk, Pkh>, NodeEvaluationState)>,
+    stack : Vec<StackElement<'stack>>,
+    altstack: Vec<StackElement<'stack>>,
+    age: &'desc u32,
+    height: &'desc u32,
```

Comment:
Why are `age` and `height` references here? Aren't these associated with a transaction/txout and have nothing to do with the descriptor?

Reply:
Yes, you are right.